Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt

```diff
@@ -0,0 +1,127 @@
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+class NetworkAvailability(private val context: Context, private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO) {
+
+    private val connectivityManager by lazy { context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager }
+    private val mutex = Mutex()
+
+    val isNetworkAvailable: Flow<Boolean> = callbackFlow {
+        val networks = mutableListOf<Network>()
+
+        val callback = object : NetworkCallback() {
+
+            override fun onAvailable(network: Network) {
+                launch {
+                    mutex.withLock {
+                        networks.add(network)
+                        sendNetworkAvailability(networks)
+                    }
+                }
+            }
+
+            override fun onLost(network: Network) {
+                launch {
+                    mutex.withLock {
+                        networks.remove(network)
+                        sendNetworkAvailability(networks)
+                    }
+                }
+            }
+        }
+
+        launch(ioDispatcher) {
+            send(getInitialNetworkAvailability(connectivityManager))
+        }
+
+        registerNetworkCallback(connectivityManager, callback)
+
+        awaitClose { unregisterNetworkCallback(connectivityManager, callback) }
+    }
+
+    private suspend fun ProducerScope<Boolean>.registerNetworkCallback(
+        connectivityManager: ConnectivityManager,
+        callback: NetworkCallback,
+    ) {
+        runCatching {
+            connectivityManager.registerNetworkCallback(networkRequestBuilder(), callback)
+        }.onFailure { exception ->
+            // Fix potential Exception thrown by ConnectivityManager on Android 11
+            // Already fixed in Android S and above
+            // https://issuetracker.google.com/issues/175055271
+            Sentry.captureException(exception)
+            send(false)
+        }
+    }
+
+    private fun unregisterNetworkCallback(connectivityManager: ConnectivityManager, callback: NetworkCallback) {
+        runCatching {
+            connectivityManager.unregisterNetworkCallback(callback)
+        }.onFailure { exception ->
+            Sentry.captureException(exception)
+        }
+    }
+
+    @Suppress("DEPRECATION")
+    private fun getInitialNetworkAvailability(connectivityManager: ConnectivityManager): Boolean {
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            connectivityManager.activeNetwork?.let(::hasInternetConnectivity) ?: false
+        } else {
+            connectivityManager.activeNetworkInfo?.isConnected ?: false
+        }
```

Comment:
```suggestion
        return connectivityManager.activeNetwork?.let(::hasInternetConnectivity) ?: false
```

Reply:
This file will be used in the Core project, which is `coreMinSdk = 21`, so I'll keep this version check.